packages/flutter_tools/lib/src/android/gradle.dart

```diff
@@ -299,3 +309,83 @@ Map<String, String> get _gradleEnv {
   }
   return env;
 }
+
+class GradleProject {
+  GradleProject(this.buildTypes, this.productFlavors, this.apkDirectory);
+
+  factory GradleProject.fromAppProperties(String properties) {
+    // Extract build directory.
+    final String buildDir = properties
+        .split('\n')
+        .firstWhere((String s) => s.startsWith('buildDir: '))
+        .substring('buildDir: '.length)
+        .trim();
+
+    // Extract build types and product flavors.
+    final Set<String> variants = new Set<String>();
+    properties.split('\n').forEach((String s) {
+      final Match match = _assembleTaskPattern.matchAsPrefix(s);
+      if (match != null) {
+        final String variant = match.group(1).toLowerCase();
+        if (!variant.endsWith('test'))
+          variants.add(variant);
+      }
+    });
+    final Set<String> buildTypes = new Set<String>();
+    final Set<String> productFlavors = new Set<String>();
+    for (final String variant1 in variants) {
+      for (final String variant2 in variants) {
+        if (variant2.startsWith(variant1) && variant2 != variant1) {
+          final String buildType = variant2.substring(variant1.length);
+          if (variants.contains(buildType)) {
+            buildTypes.add(buildType);
+            productFlavors.add(variant1);
+          }
+        }
+      }
+    }
+    if (productFlavors.isEmpty)
+      buildTypes.addAll(variants);
+    return new GradleProject(
+      buildTypes.toList(),
+      productFlavors.toList(),
+      fs.path.join(buildDir, 'outputs', 'apk'),
+    );
+  }
+
+  final List<String> buildTypes;
+  final List<String> productFlavors;
```

Comment:
Yes. Lists are easier to create (for testing, defaults); there is no Set literal yet. And the expected length of these lists (2-4) is small enough that they are likely to outperform sets, even for `contains`.